# 内存回收

## 回收那些空间
1. 在运行中，程序计数器、虚拟机栈、本地方法栈随着想成的创建而创建，也随线程的消亡而消亡，并且方法的栈帧 基本上是确定大小的，在类结构确定后 大小基本确定
2. 整个垃圾回收主要的区域是堆和方法区

## 对象已死

### 引用计数
在对象中添加一个引用计数器，当对象被引用时候增加1，对象不被引用时候减少1，当对象不再引用就可以代表回收了。但是在主流的JVM 中并不真正的使用到该算法<br/>
比如对象ObjA，ObjB 中都拥有属性 instance 指向 对方，其实这两个不再被使用，但依旧回收不了
### 可达性分析法
通过被称为"GC Roots"的对象做为起点集，从这些点开始根据引用关系向下搜索，搜索走过的路径成为引用链（Reference Chain）,如果某个对象与GC Roots 没有任何的引用链 或者说从GC Roots 到该对象不可达，则证明该对象不再被使用

#### GC Roots
- 虚拟机栈中引用的对象，比如当前运行的参数，局部变量，临时变量
- 方法区中静态属性引用的对象
- 方法区中常量引用的对象，比如字符串常量
- JNI 中引用的变量
- 虚拟机内部引用，比如Class 对象，异常信息，类加载器
- 同步锁（syschronized） 持有对象
- 反映Java 虚拟机内存情况的JMXBean,本地代码缓存等

除了固定的GC Roots 外还有一些根据垃圾回收器和回收区域临时的性加入的对象，共同组建成GC Roots


## 引用
JDK1.2 定义：如果引用类型数据中存储的数值代表的是另外一块内存的地址或起始地址，则成该refernce 类型代表了某块内存、某个对象的引用。这种表示关系太过狭隘，一个对象不仅只有引用和未引用的关系，所以扩充了引用的概念，强度依次减弱

### 强引用 （Strongly Reference ）
最传统的引用关系，指的是代码之间的普遍的引用赋指，只要强引用关系存在，垃圾回收器一定不会回收该对象
### 软引用 (Soft Reference)
描述一些还有用但是非必须的对象，JDK 提供SoftReference 类来描述软引用，只要软引用关系存在，在系统即将发生内存溢出前，才会把这些对象列进后手范围，如果回收之后内存依旧不够，则抛出内存溢出异常
### 弱引用 (Week Reference)
表述非必须的对象，JDK 提供WeekReference 类来描述弱引用，弱引用对象只能生存在下一次垃圾回收器工作发生为止，当垃圾回收器开始工作时候，无论内存是否足够都会回收弱引用关联的对象
### 虚引用 (Phantom Reference)
最弱的引用关系，JDK 提供 PhantomReference 类来表述，为一个对象设置引用关系的目的是为了能在这个对象被垃圾回收器回收前收到系统通知


## 回收？
即使可达性分析算法判定为不可达对象，并不一定会回收该对象，该对象处于"缓刑"状态，真正进行回收一个对象最多经过两个标记过程：
1. 进行可达性分析算法后，发现并没有GC Roots 有连接的引用链，进行第一次标记 判读对象是否有必要执行finalize 方法，假设对象没有覆盖finalize 方法，或者finalize 方法已经被虚拟机调用过，那么认为没有必要执行
<br/>
如果该对象被判定没有必要执行finalize 方法，那么该对象将会被放置在一个名称为F-Queue 的队列中，稍后由一条虚拟机建立，低调度优先级的Finalizer 线程去执行他们的finalize 方法，这里的执行代表的仅仅是触发，并不会等待执行结束
2. 对F-Queue 进行第二次标记，如果类在finalize 方法中自救，比如自己赋值给某个类变量或者对象的成员变量，那么第二次标记会将它移除队列，如果没有自救则真正回收
```java
package tets;
public class FinalizeEscapeGC {

    public static FinalizeEscapeGC SAVE_HOOK = null;

    public void isActive() {
        System.out.println("I`m alive");
    }

    @Override
    protected void finalize() throws Throwable {
        super.finalize();
        System.out.println("finalize exec");
        SAVE_HOOK = this; // 对象进行自我拯救
    }

    public static void main(String[] args) throws InterruptedException {
        SAVE_HOOK = new FinalizeEscapeGC();
        SAVE_HOOK = null;
        System.gc();
        Thread.sleep(500);

        if (SAVE_HOOK != null) {
            SAVE_HOOK.isActive();
        } else {
            System.out.println("dead");
        }

    }
}
```

执行出
```
finalize exec
I`m alive
```

可以看到确实进行了对象的自我拯救，但是如果main 执行两次，对象将无法自我拯救,<br/>
因为任何一个对象的finalize()方法只会被虚拟机调用一次，如果对象面临下一次回收，他的finalize 方法不会被再次执行